src/WinRT.Runtime2/InteropServices/ExceptionHelpers.cs

```diff
@@ -0,0 +1,264 @@
+    // manage the lifetime of the CCW for the exception object to avoid cycles and thereby leaking it.
+    internal static unsafe Exception? GetLanguageException(void* languageErrorInfoPtr, int hr)
+    {
+        // Check the error info first for the language exception.
+        Exception? exception = GetLanguageExceptionInternal(languageErrorInfoPtr, hr);
+        if (exception is not null)
+        {
+            return exception;
+        }
+
+        // If propagated exceptions are supported, traverse it and check if any one of those is our exception to reuse.
+        if (Marshal.QueryInterface((nint)languageErrorInfoPtr, WellKnownInterfaceIds.IID_ILanguageExceptionErrorInfo2, out nint languageErrorInfo2Ptr) >= 0)
+        {
+            void* currentLanguageExceptionErrorInfo2Ptr = null;
+            try
+            {
+                if (WellKnownErrorCodes.Succeeded(ILanguageExceptionErrorInfo2Vftbl.GetPropagationContextHeadUnsafe((void*)languageErrorInfo2Ptr, &currentLanguageExceptionErrorInfo2Ptr)))
+                {
+                    while (currentLanguageExceptionErrorInfo2Ptr != null)
+                    {
+                        Exception? propagatedException = GetLanguageExceptionInternal(currentLanguageExceptionErrorInfo2Ptr, hr);
+                        if (propagatedException is not null)
+                        {
+                            return propagatedException;
+                        }
+
+                        void* previousLanguageExceptionErrorInfo2Ptr = currentLanguageExceptionErrorInfo2Ptr;
+                        try
+                        {
+                            _ = ILanguageExceptionErrorInfo2Vftbl.GetPreviousLanguageExceptionErrorInfoUnsafe(currentLanguageExceptionErrorInfo2Ptr, &currentLanguageExceptionErrorInfo2Ptr);
+                        }
+                        finally
+                        {
+                            _ = Marshal.Release((nint)previousLanguageExceptionErrorInfo2Ptr);
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                MarshalExtensions.ReleaseIfNotNull(currentLanguageExceptionErrorInfo2Ptr);
+                _ = Marshal.Release(languageErrorInfo2Ptr);
+            }
+        }
+
+        return null;
+    }
+
+    private static unsafe Exception? GetLanguageExceptionInternal(void* languageErrorInfoPtr, int hr)
+    {
+        void* languageExceptionPtr = null;
+        try
+        {
+            if (WellKnownErrorCodes.Succeeded(ILanguageExceptionErrorInfoVftbl.GetLanguageExceptionUnsafe(languageErrorInfoPtr, &languageExceptionPtr)) && languageExceptionPtr != null)
+            {
+                if (WindowsRuntimeMarshal.IsReferenceToManagedObject(languageExceptionPtr))
+                {
+                    Exception ex = ComInterfaceDispatch.GetInstance<Exception>((ComInterfaceDispatch*)languageExceptionPtr);
```

Comment:
You can simplify this by replacing the `null` check + `IsReferenceToManagedObject` + `ComInterfaceDispatch` call with just a single call to `WindowsRuntimeMarshal.TryGetManagedObject`, it does exactly the same as all this 🙂